### ISSUE TITLE
[json reporter]: Report suites/tests that have not started yet

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/test#1654]
## Problem 

Currently, the json reporter does not give informations about tests/groups/suites unless they start to execute

Meaning that if one particular test takes a lot of time to complete, for the entire time the renderer wouldn't be able to know that more tests are yet to come.

## Proposal

One solution could be to add an event similar to "test start", but would be sent the moment the `test()` function is called instead of when the test handler (`test('...', handler)`) starts

And we'd have the same for suites/groups

Meaning that if we had:

```dart
group('group', () {
  test('slow', () => Future.delayed(Duration(seconds: 2)));

  test('another, () {});
});
```

then we would receive in order:

```
{ "type": "group", "group": {...} }
{ "type": "test", "test": {"name": "slow", ...} }
{ "type": "test", "test": {"name": "another",...} }
{ "type": "test_start", "test": {"name": "slow",...}}
{ "type": "test_end", "test": {"name": "slow",...}}
{ "type": "test_start", "test": {"name": "another",...}}
{ "type": "test_end", "test": {"name": "another",...}}
```

**Comment by jakemac53** [comment]
Unfortunately the design of this package doesn't really support this well. We can't reliably discover tests until the test suite is actually loaded and its `main` executed.

We could choose to more eagerly load and execute `main`, but the problem is that users can and do have arbitrary code in `main`, which might be very slow. They should really use a `setupAll` for anything slow, but don't necessarily, as there isn't a ton of user-visible difference between doing that and setting things up in `main` .

**Comment by jakemac53** [comment]
We could expose the list of `suites` earlier potentially, before we even attempt to load them? But that won't give you the list of tests within the suite.

**Comment by DanTup** [comment]
> We could expose the list of `suites` earlier potentially, before we even attempt to load them? But that won't give you the list of tests within the suite.

I've often wanted this - I think it would at least allow me to tell VS Code that those files are starting (so that the user would see them change to spinners or something right away, giving a more obvious sense of the progress).

**Comment by rrousselGit** [comment]
I don't mind the lazy loading of groups/tests

At least being able to get the list of all suites from the get go would be nice.

This can allow rendering a progress indicator like "x suites out of y completed"

**Comment by jakemac53** [comment]
That all seems reasonable to me, we may want to do a separate event though from the loading event, as the suite isn't actually loading yet?

**Comment by rrousselGit** [comment]
I'm fine with that. That's how I expected it'd be

**Comment by DanTup** [comment]
sgtm too - perhaps a "discovered" event?

At the risk of inventing more work, is the same for groups/tests? Do we only find out about them when they actually start executing? If so, would it make sense to have equiv discovery events for suites/tests (where those discovered prior to executing starting could also be sent - allowing the editor to also give the same fine-grain progress of tests within)?

**Comment by jakemac53** [comment]
Discovered event SGTM. Not sure if it makes sense for groups/tests or not, the time between discovery and actual running of those should usually be very small, unless you are doing really weird things in your `main` like defining some test, then awaiting some long async task, then defining more tests.

**Comment by DanTup** [comment]
> Not sure if it makes sense for groups/tests or not, the time between discovery and actual running of those should usually be very small, unless you are doing really weird things in your `main` like defining some test, then awaiting some long async task, then defining more tests.

What about just a slow test?

```dart
import 'package:test/test.dart';

main() {
  group('foo', () {
    test('test 1', () async {
      print('test 1 starting...');
      await Future.delayed(const Duration(seconds: 5));
      print('test 1 ending...');
      expect(1, 1);
    });
    test('test 2', () async {
      print('test 2 starting...');
      await Future.delayed(const Duration(seconds: 5));
      print('test 2 ending...');
      expect(1, 1);
    });
  });
}
```

Here, I don't see any mention of `test 2` in the JSON output until test 1 completes, but `package:test` presumably knows about `test 2` before any test started running (I presume it already ran `main` to completion before it invoked `test 1`)?

**Comment by jakemac53** [comment]
> Here, I don't see any mention of `test 2` in the JSON output until test 1 completes, but `package:test` presumably knows about `test 2` before any test started running

Oh ok then yes in that case yes it makes sense to get some sort of earlier event.

Do we want to send an event for each test/group as it is discovered or a single one after main is done executing with all the tests/groups?

> (I presume it already ran `main` to completion before it invoked `test 1`)?

Yes

**Comment by DanTup** [comment]
> Do we want to send an event for each test/group as it is discovered or a single one after main is done executing with all the tests/groups?

Is it possible for users to register tests after main completed? If so, individual events probably make sense (since some may arrive after tests start running). Otherwise, a single one might be better 🤷‍♂️  although I don't think it'd make a real difference for me/VS Code.

**Comment by jakemac53** [comment]
> Is it possible for users to register tests after main completed?

No, it is not. They can return a Future from main and declare tests until that completes, but after that any new tests aren't respected. I think you will get an error if you try to define more, but I can't recall for sure.

**Comment by DanTup** [comment]
Ah, got it - no preference from me then. Any advantage or either I can think of is very minor at best.
